File: engine/engine-model/src/main/scala/client/database/data/WorldRankData.scala
package client.database.data

import scala.jdk.CollectionConverters._

class WorldRankData(private var _worldId: Int) {
  def worldId: Int = _worldId

  private var _userRanks: List[GlobalUserRank] = List[GlobalUserRank]()

  def userRanks: List[GlobalUserRank] = _userRanks

  def addUserRank(userRank: GlobalUserRank): Unit = {
    _userRanks = _userRanks.appended(userRank)
  }

  def getUserRanks: java.util.List[GlobalUserRank] = {
    _userRanks.asJava
  }
}

File: engine/engine-model/src/main/java/client/database/data/WorldRankData.java
package client.database.data;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public record WorldRankData(Integer worldId, List<GlobalUserRank> userRanks) {
   public WorldRankData(Integer worldId) {
      this(worldId, Collections.emptyList());
   }

   public WorldRankData addUserRank(GlobalUserRank userRank) {
      return new WorldRankData(worldId, Stream.concat(userRanks.stream(), Stream.of(userRank)).collect(Collectors.toUnmodifiableList()));
   }
}
